### ISSUE TITLE
Subscription challenges

### ISSUE THREAD

**Issue by michael-woolsey** [gql-dart/ferry#505]
Hello,

I'm attempting to get subscriptions working with Ferry and I'm encountering challenges that are very hard for me to debug.  I found issue (https://github.com/gql-dart/ferry/issues/387) that's been closed stating to look at various `graphql` documents.

However these documents appear to be using variables and methods that are not available to me in Ferry.  

Such as:
1. `Link.split` -> It's supposed to have `request.isSubscription` but the object I'm working with does not have this variable.  I've tried working around the issue by utilizing my operation name so this isn't critical yet.
2. These docs also state to use `client.subscribe(...)` however the Ferry client doesn't appear to have such a method.

I know my server is working with subscriptions as I can use the `graphiql` webpage and when I update my user name the subscription reflects the change in the webpage immediately.

I'm attempting a simple text widget to see the users name change but it's not showing any data, it's always coming back as `null`.

I'm splitting the link as follows:
```
Link.split(
  (request) {
    return request.operation.operationName?.toLowerCase().endsWith('subscription') ?? false;
  },
  WebSocketLink(Environment.wsEndpoint),
  HttpLink(params.apiUrl, httpClient: networkClient),
),
```

This is the widget I'm trying to use to view the name changes:
```
class NameCard extends StatelessWidget {
  const NameCard({super.key});

  @override
  Widget build(BuildContext context) {
    return Operation(
      client: app<FerryClient>(),
      operationRequest: GNameUpdatedSubscriptionReq(),
      builder: (
        BuildContext context,
        OperationResponse<GNameUpdatedSubscriptionData, GNameUpdatedSubscriptionVars>? response,
        Object? error,
      ) {
        if (response != null && response.hasErrors) {
          throw response.linkException ?? response.graphqlErrors!.first;
        }
        return Text(response?.data?.currentUserUpdated?.user?.name ?? 'N/A');
      },
    );
  }
}
```

__edit__
This is my graphql query:
```
subscription NameUpdatedSubscription {
  currentUserUpdated {
    event
    user {
      name
    }
  }
}
```

I want to also mention that under normal query circumstances, my queries all work as expected.  So I know my server is capable of sending me the data and my app is capable of receiving it.

However, with subscriptions I'm not getting any error messages, I tried putting a completely bogus websocket address and that caused an exception, so I know it's finding the server.  However when I receive nothing but `null` values and no error messages I'm not entirely sure how to proceed.

Lastly, I should mention that we're using the `IsolateClient`, just in case that makes a difference.

Does anyone have any suggestions as having subscriptions work it becoming more and more important.

Thank you,
Michael

**Comment by knaeckeKami** [comment]
Hi,

in the latest alpha version of gql_exec there is the [getOperationType](https://github.com/gql-dart/gql/blob/master/links/gql_exec/lib/src/operation_operationtype_extension.dart)() extension on the Operation type, which should allow to easily to determine which operation is a Subscription.


With ferry, you can just use client.request(), since ferry is stream based anyway the API for queries and subscriptions is the same. 

It might be that your server only supports the newer graphql-transport-ws protocol, not the legacy subscriptions-transport-ws protocol which is used by WebsocketLink.

There is a pull request for the new PR here: https://github.com/gql-dart/gql/pull/355
and graphql-flutter also has an implementation of a Websocket link that supports the new protocol here: https://github.com/zino-hofmann/graphql-flutter/blob/main/packages/graphql/lib/src/links/websocket_link/websocket_link.dart , but it's not part of gql-dart unfortunately (but is uses the same Link base class and should be compatible)

**Comment by michael-woolsey** [comment]
Thank you @knaeckeKami -- I'll take a look at this information shortly!

**Comment by knaeckeKami** [comment]
Another thing that I forgot: since you are using the IsolateClient, be sure to pass the onError SendPort in case there are uncaught exceptions on the other Isolate.

You can obtain the SendPort by creating a ReceivePort and then accessing its sendPort. Then listen to the ReceivePort for mesages.

**Comment by michael-woolsey** [comment]
I discovered the root problem I was having with this issue... it came down to two things:
1. I had to use the code from https://github.com/gql-dart/gql/pull/355, most likely due to needing the `graphql-transport-ws` protocol.
2. I had to update some code to pass in headers, I realize I could have passed in a socket maker generator, but this is the approach I took.

I changed
```
        socket = WebSocketChannel.connect(
          Uri.parse(await options.socketMaker.url!()),
          protocols: [graphQLTransportWSProtocol],
        );
```
To be this:
```
        socket = IOWebSocketChannel.connect(
          Uri.parse(await options.socketMaker.url!()),
          protocols: [graphQLTransportWSProtocol],
          headers: options.headers,
        );
```

Note that I had to use `IOWebSocketChannel` instead of `WebSocketChannel` to expose the required `headers` property using the `import 'package:web_socket_channel/io.dart';` import.

I also had to update the `ClientOptions` class to have a property called `final Map<String, dynamic>? headers;` so that I could access the variable I needed.

Hopefully this helps someone else should they have similar challenges attempting to get subscriptions running using cookie based authentication.

Michael
